pysecs/secs.py

```diff
@@ -102,40 +102,52 @@ def fit(self, obs_loc, obs_B, obs_var=None, epsilon=0.05):
         if obs_loc.shape[-1] != 3:
             raise ValueError("Observation locations must have 3 columns (lat, lon, r)")
 
-        # Assume unit variance of all measurements
-        if obs_var is None:
-            obs_var = np.ones(obs_loc.shape)
-
         if obs_B.ndim == 2:
             # Just a single snapshot given, so expand the dimensionality
             obs_B = obs_B[np.newaxis, ...]
 
+        # Assume unit variance of all measurements
+        if obs_var is None:
+            obs_var = np.ones(obs_B.shape)
+
         ntimes = len(obs_B)
 
         # Calculate the transfer functions
         T_obs = self._calc_T(obs_loc)
 
-        # Calculate the singular value decomposition (SVD)
-
-        # T has shape: (nobs, 3, nsec),
-        # flatten the directions and reshape it to (nobs*3, nsec)
-        U, S, Vh = np.linalg.svd(T_obs.reshape(-1, self.nsec) /
-                                 obs_var.ravel()[:, np.newaxis], full_matrices=False)
-
-        # Divide by infinity (1/S) gives zero weights
-        W = 1./S
-        # Eliminate the small singular values (less than epsilon)
-        # by giving them zero weight
-        W[S < epsilon*S.max()] = 0.
+        # Store the fit sec_amps in the object
+        self.sec_amps = np.empty((ntimes, self.nsec))
+        self.sec_amps_var = np.empty((ntimes, self.nsec))
 
-        VWU = np.dot(Vh.T, np.dot(np.diag(W), U.T))
-
-        # Store the fit 'sec_amps' in the object
-        # shape: ntimes x nsec
-        self.sec_amps = np.dot(VWU, (obs_B/obs_var).reshape(ntimes, -1).T).T
-        # Maybe want the variance of the predictions sometime later...?
-        # shape: nsec
-        self.sec_amps_var = np.sum((Vh.T * W)**2, axis=1)
+        # Calculate the singular value decomposition (SVD)
+        # T_obs has shape (nobs, 3, nsec), we need to reshape it
+        # and add an extra dimension for time, making it (1, nobs*3, nsec)
+        # obs_var has shape (ntimes, nobs, 3), reshape it and add an
+        # extra dimension for the secs, making it (ntimes, nobs*3, 1)
+
+        # After all of the reshaping and division, the full svd_in matrix
+        # has shape (ntimes, nobs*3, nsec)
+        svd_in = (T_obs.reshape(-1, self.nsec)[np.newaxis, :] /
+                  obs_var.reshape((ntimes, -1))[..., np.newaxis])
+        for i in range(ntimes):
+            # Iterate through all times, only calculating the SVD when needed
+            # The first timestep, or anytime the variance changes
+            if i == 0 or not np.all(svd_in[i, ...] == svd_in[i-1, ...]):
```

Comment:
I suspect it's not much, but this element-by-element comparison could get a little burdensome. In my implementation, I actually check to see if the object being referenced had changed (in other words, I use `X is Y` instead of `np.all(X == Y)`. Element-by-element array comparisons in NumPy are pretty fast though, so my solution is probably overkill, and not all that intuitive, TBH. I just bring it up here in case you were interested in (probably slight) performance improvements.

Reply:
In fact, I just checked more closely, and the actual SVD call takes far more time than the element-by-element comparison of arrays, even with a relatively small matrix. Comparing objects, instead of comparing arrays element-by-element, will not result in a noticeable performance improvement.

Reply:
Unfortunately, the way it is implemented with `arr[i, ...] is arr[i-1, ...]` wouldn't work :( since they are not explicitly the same object :( So, it would require a refactor to store all of the arrays in the SECS object itself more like yours. I think we'll wait on that to do in a later version.